### PR TITLE
docs: add the HTTP request documentation. Improve features' docs

### LINKED ISSUE #179: Update the documentation to include information and examples about the HTTP bindings
### Is your feature request related to a problem? Please describe.

The new `HTTP` bindings are not documented in Wasm Workers Server. Developers may not know the new feature is available.

### Describe the solution you'd like

I want to update the documentation to include the following information:

- [x] A new page for this feature
- [x] Update the "Features" to show a compatibility table
- [x] Include a compatibility table in all the "language" pages
- [x] Add examples for the different supported languages

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Angelmmiguel**
Add the new documentation about the HTTP bindings. Since this feature is not available yet in all the languages, I also added "compatibility" tables in different pages to ensure the language support is clear. I also updated the "Features" main page to add more information and a global compatibility table.

It closes #179 

## Changes

* Improve the "Features" page with more information
* Add a new page about HTTP requests
* Add HTTP request examples in Go, Rust and JavaScript
* Add compatibility tables to different pages

## Preview

https://http-requests.wasm-workers-docs.pages.dev/

### CODE REVIEW

File: docs/docs/features/all.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# Features
+
+Wasm Workers Server provides different features to develop serverless applications. Some of these features are related to the server like the static asset management, while others relate to the workers like sending HTTP requests.
```
Comment (ereslibre) [inline docs/docs/features/all.md:3]:
```suggestion
Wasm Workers Server provides different features to develop serverless applications. Some of these features are related to the server like the static asset management, while others relate to workers like sending HTTP requests.
```

File: docs/docs/features/all.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+* **Server features**: customizes or expands Wasm Worker Server capabilities. For example, you can expose static assets by saving these files in a `public` folder.
+
+* **Worker features**: expose new features to the individual workers so they can perform more complex tasks. For example, workers can access to a K/V store or use environment variables.
+
+## Available features
+
+### Server
+
+* [Static assets' management](./static-assets.md)
```
Comment (ereslibre) [inline docs/docs/features/all.md:15]:
```suggestion
* [Static assets management](./static-assets.md)
```

File: docs/docs/features/all.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+* [Multiple language runtimes](./multiple-language-runtimes.md)
+
+### Workers
+
+* [Key / Value store](./key-value.md)
+* [HTTP Requests (fetch)](./http-requests.md)
+* [Dynamic routes](./dynamic-routes.md)
+* [Environment variables](./environment-variables.md)
+* [Mount folders](./environment-variables.md)
```
Comment (ereslibre) [inline docs/docs/features/all.md:24]:
```suggestion
* [Mount folders](./mount-folders.md)
```

File: docs/docs/features/all.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+## Language compatibility
+
+You can develop workers in different languages. However, not all of them support all features. **The goal is to support all of them**, although there are some constraints that makes some features more complex to implement in certain languages.
```
Comment (ereslibre) [inline docs/docs/features/all.md:28]:
```suggestion
You can develop workers in different languages. However, not all of them support all features. **The goal is to support all of them**, although there are some constraints that make some features more complex to implement in certain languages.
```

File: docs/docs/features/all.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+The following table shows the language compatibility for the different worker's functions:
```
Comment (ereslibre) [inline docs/docs/features/all.md:30]:
```suggestion
The following table shows the language compatibility for the different worker functions:
```

File: docs/docs/features/http-requests.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
+---
+sidebar_position: 3
+---
+
+# HTTP Requests (fetch)
+
+Many times, workers require to access data from an external resource like a website or an API. This feature allows workers to perform HTTP requests to external resources from a worker. It follows the capability-based model, so workers cannot perform any HTTP request until you configure the allowed hosts and HTTP methods.
```
Comment (ereslibre) [inline docs/docs/features/http-requests.md:7]:
```suggestion
Often times, workers require to access data from an external resource like a website or an API. This feature allows workers to perform HTTP requests to external resources. It follows the capability-based model, so workers cannot perform any HTTP request until you configure the allowed hosts and HTTP methods.
```

File: docs/docs/features/http-requests.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+[features.http_requests]
+allowed_methods = ["GET", "POST"]
+allowed_hosts = ["jsonplaceholder.typicode.com"]
+```
+
+Now, your worker can perform HTTP requests following those rules.
+
+## Send HTTP requests in different languages
+
+Depending on the language, the different kits exposes this feature in a different way. The goal is to use a common API to perform HTTP requests in that language. For example, to perform HTTP requests in JavaScript you can use the [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) API.
```
Comment (ereslibre) [inline docs/docs/features/http-requests.md:25]:
```suggestion
Depending on the language, the different kits expose this feature in a different way. The goal is to use a common API to perform HTTP requests in that language. For example, to perform HTTP requests in JavaScript you can use the [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) API.
```

File: docs/docs/languages/go.md [hunk 8]
```diff
@@ -234,6 +234,127 @@ To add a KV store to your worker, follow these steps:
 
 1. Finally, open <http://127.0.0.1:8080/worker-kv> in your browser.
 
+## Send an HTTP request
+
+Wasm Workers allows you to send HTTP requests from your workers. Read more information about this feature in the [HTTP Requests](../features/http-requests.md) section.
+
+To perform a HTTP requests from your worker, follow these steps:
```
Comment (ereslibre) [inline docs/docs/languages/go.md:241]:
```suggestion
To perform HTTP requests from your worker, follow these steps:
```

File: docs/docs/languages/go.md [hunk 9]
```diff
@@ -234,6 +234,127 @@ To add a KV store to your worker, follow these steps:
...
+1. Create a `fetch.go` file with the following contents:
+
+    ```go title="fetch.go"
+    package main
+
+    import (
+        "io"
+        "net/http"
+
+	    "github.com/vmware-labs/wasm-workers-server/kits/go/worker"
```
Comment (ereslibre) [inline docs/docs/languages/go.md:264]: This one looks off

File: docs/docs/languages/go.md [hunk 10]
```diff
@@ -234,6 +234,127 @@ To add a KV store to your worker, follow these steps:
...
+1. Then, let's create the `http.Request` instance and pass it to the `worker.SendHttpRequest` method. In this example, we will call the [{JSON} Placeholder API](https://jsonplaceholder.typicode.com/) to retrieve a `Post`. You will read the content of the response using the `gjson` API:
+
+    ```go title="fetch.go"
+    package main
+
+    import (
+        "io"
+        "net/http"
+
+	    "github.com/vmware-labs/wasm-workers-server/kits/go/worker"
```
Comment (ereslibre) [inline docs/docs/languages/go.md:286]: Same, this one looks off

File: docs/docs/languages/go.md [hunk 11]
```diff
@@ -234,6 +234,127 @@ To add a KV store to your worker, follow these steps:
...
+    }
+    ```
+
+1. Compile the project to Wasm ([WASI](https://wasi.dev/)):
+
+    ```bash
+    tinygo build -o fetch.wasm -target wasi fetch.go
+    ```
+
+1. Create a `fetch.toml` file with the following content. It enables the worker to perform the HTTP request to that host. By default, HTTP requests are forbidden.
```
Comment (ereslibre) [inline docs/docs/languages/go.md:328]:
```suggestion
1. Create a `fetch.toml` file with the following content. It enables the worker to perform HTTP requests to that host given that, by default, HTTP requests are forbidden.
```

File: docs/docs/languages/go.md [hunk 12]
```diff
@@ -234,6 +234,127 @@ To add a KV store to your worker, follow these steps:
...
+
+    [features]
+    [features.http_requests]
+    allowed_methods = ["GET"]
+    allowed_hosts = ["jsonplaceholder.typicode.com"]
+    ```
+
+1. Run your worker with `wws`. If you didn't download the `wws` server yet, check our [Getting Started](../get-started/quickstart.md) guide.
+
+    ```bash
```
Comment (ereslibre) [inline docs/docs/languages/go.md:344]:
```suggestion
    ```shell-session
```

File: docs/docs/languages/javascript.md [hunk 13]
```diff
@@ -137,6 +137,79 @@ To add a KV store to your worker, follow these steps:
 
 1. Finally, open <http://127.0.0.1:8080/counter> in your browser.
 
+## Send an HTTP request (fetch)
+
+Wasm Workers allows you to send HTTP requests from your workers. Read more information about this feature in the [HTTP Requests](../features/http-requests.md) section.
+
+To perform a HTTP requests from your worker, follow these steps:
```
Comment (ereslibre) [inline docs/docs/languages/javascript.md:144]:
```suggestion
To perform HTTP requests from your worker, follow these steps:
```

File: docs/docs/languages/javascript.md [hunk 14]
```diff
@@ -137,6 +137,79 @@ To add a KV store to your worker, follow these steps:
...
+
+        return new Response(body);
+    }
+
+    addEventListener("fetch", event => {
+        event.respondWith(reply(event.request));
+    });
+    ```
+
+1. Create an `index.toml` file with the following content. It enables the worker to perform the HTTP request to that host. By default, HTTP requests are forbidden.
```
Comment (ereslibre) [inline docs/docs/languages/javascript.md:183]:
```suggestion
1. Create an `index.toml` file with the following content. It enables the worker to perform HTTP requests to that host given that, by default, HTTP requests are forbidden.
```

File: docs/docs/languages/javascript.md [hunk 15]
```diff
@@ -137,6 +137,79 @@ To add a KV store to your worker, follow these steps:
...
+
+    [features]
+    [features.http_requests]
+    allowed_methods = ["GET"]
+    allowed_hosts = ["jsonplaceholder.typicode.com"]
+    ```
+
+1. Save the file and run your worker with `wws`. If you didn't download the `wws` server yet, check our [Getting Started](../get-started/quickstart.md) guide.
+
+    ```bash
```
Comment (ereslibre) [inline docs/docs/languages/javascript.md:199]:
```suggestion
    ```shell-session
```

File: docs/docs/languages/rust.md [hunk 16]
```diff
@@ -231,6 +231,156 @@ To add a KV store to your worker, follow these steps:
 
 1. Finally, open <http://127.0.0.1:8080/worker-kv> in your browser.
 
+## Send an HTTP request
+
+Wasm Workers allows you to send HTTP requests from your workers. Read more information about this feature in the [HTTP Requests](../features/http-requests.md) section.
+
+To perform a HTTP requests from your worker, follow these steps:
```
Comment (ereslibre) [inline docs/docs/languages/rust.md:238]:
```suggestion
To perform HTTP requests from your worker, follow these steps:
```

File: docs/docs/languages/rust.md [hunk 17]
```diff
@@ -231,6 +231,156 @@ To add a KV store to your worker, follow these steps:
...
+
+1. Create a new Rust project:
+
+    ```bash
+    cargo new --name fetch fetch
+    ```
+
+1. Add the dependencies to the `Cargo.toml` file:
```
Comment (ereslibre) [inline docs/docs/languages/rust.md:246]:
```suggestion
1. Add dependencies to the `Cargo.toml` file:
```

File: docs/docs/languages/rust.md [hunk 18]
```diff
@@ -231,6 +231,156 @@ To add a KV store to your worker, follow these steps:
...
+        // Get the request
+        let res = bindings::send_http_request(external_request).unwrap();
+
+        // Parse the response
+        let data = res.body();
+
+        let post: Post = serde_json::from_slice(&data).unwrap();
+
+        // Applied changes here to use the Response method. This requires changes
+        // on signature and how it returns the data.
```
Comment (ereslibre) [inline docs/docs/languages/rust.md:318]: I this meant to be here?

File: docs/docs/languages/rust.md [hunk 19]
```diff
@@ -231,6 +231,156 @@ To add a KV store to your worker, follow these steps:
...
+      cargo build --release --target wasm32-wasi
+    ```
+
+1. After you compiled the project, move the worker to the current folder:
+
+    ```bash
+    mv ./target/wasm32-wasi/release/fetch.wasm ./
+    ```
+
+1. Create a `fetch.toml` file with the following content. It enables the worker to perform the HTTP request to that host. By default, HTTP requests are forbidden.
```
Comment (ereslibre) [inline docs/docs/languages/rust.md:356]:
```suggestion
1. Create a `fetch.toml` file with the following content. It enables the worker to perform HTTP requests to that host given that, by default, HTTP requests are forbidden.
```

File: docs/docs/languages/rust.md [hunk 20]
```diff
@@ -231,6 +231,156 @@ To add a KV store to your worker, follow these steps:
...
+
+    [features]
+    [features.http_requests]
+    allowed_methods = ["GET"]
+    allowed_hosts = ["jsonplaceholder.typicode.com"]
+    ```
+
+1. Run your worker with `wws`. If you didn't download the `wws` server yet, check our [Getting Started](../get-started/quickstart.md) guide.
+
+    ```bash
```
Comment (ereslibre) [inline docs/docs/languages/rust.md:372]:
```suggestion
    ```shell-session
```

File: docs/docs/languages/rust.md [hunk 21]
```diff
@@ -231,6 +231,156 @@ To add a KV store to your worker, follow these steps:
...
 
+## Send an HTTP request
+
+Wasm Workers allows you to send HTTP requests from your workers. Read more information about this feature in the [HTTP Requests](../features/http-requests.md) section.
+
+To perform a HTTP requests from your worker, follow these steps:
+
+1. Create a new Rust project:
+
+    ```bash
```
Comment (ereslibre) [inline docs/docs/languages/rust.md:242]:
```suggestion
    ```shell-session
```

File: docs/docs/languages/go.md [hunk 22]
```diff
@@ -234,6 +234,127 @@ To add a KV store to your worker, follow these steps:
...
 
+## Send an HTTP request
+
+Wasm Workers allows you to send HTTP requests from your workers. Read more information about this feature in the [HTTP Requests](../features/http-requests.md) section.
+
+To perform a HTTP requests from your worker, follow these steps:
+
+1. Create a new Go project:
+
+    ```bash
```
Comment (ereslibre) [inline docs/docs/languages/go.md:245]:
```suggestion
    ```shell-session
```

File: docs/docs/languages/go.md [hunk 23]
```diff
@@ -234,6 +234,127 @@ To add a KV store to your worker, follow these steps:
...
+
+            w.Header().Set("x-generated-by", "wasm-workers-server")
+            w.Write([]byte(title.String()))
+        })
+    }
+    ```
+
+1. Compile the project to Wasm ([WASI](https://wasi.dev/)):
+
+    ```bash
```
Comment (ereslibre) [inline docs/docs/languages/go.md:324]:
```suggestion
    ```shell-session
```

File: docs/docs/languages/rust.md [hunk 24]
```diff
@@ -231,6 +231,156 @@ To add a KV store to your worker, follow these steps:
...
+        Ok(http::Response::builder()
+            .status(200)
+            .header("x-generated-by", "wasm-workers-server")
+            .body(response.into())?)
+    }
+    ```
+
+1. Compile the project to Wasm ([WASI](https://wasi.dev/)):
+
+    ```bash
```
Comment (ereslibre) [inline docs/docs/languages/rust.md:344]:
```suggestion
    ```shell-session
```

File: docs/docs/languages/rust.md [hunk 25]
```diff
@@ -231,6 +231,156 @@ To add a KV store to your worker, follow these steps:
...
+    # Install the component and build
+    rustup target add wasm32-wasi && \
+      cargo build --release --target wasm32-wasi
+    ```
+
+1. After you compiled the project, move the worker to the current folder:
+
+    ```bash
```
Comment (ereslibre) [inline docs/docs/languages/rust.md:352]:
```suggestion
    ```shell-session
```

**Review by ereslibre — COMMENTED** [review]
Looks great! Some comments

File: docs/docs/languages/go.md (hunk 9 above)
Comment (Angelmmiguel) [reply docs/docs/languages/go.md:264]: What do you mean? 🤔

File: docs/docs/languages/rust.md (hunk 18 above)
Comment (Angelmmiguel) [reply docs/docs/languages/rust.md:318]: Nooop! I'm gonna remove it hehe

File: docs/docs/languages/go.md (hunk 9 above)
Comment (ereslibre) [reply docs/docs/languages/go.md:264]: The indentation :)

Comment (Angelmmiguel) [reply docs/docs/languages/go.md:264]: Ahhh! I think this case is correct as these libraries are arguments for the previous `import`. Maybe I'm missing something as I'm not very familiar with Go 🤔.

This is another example that uses 2 space indentation. The code block in the documentation uses 4.

```go
package main

import (
	"bytes"
	"io"
	"net/http"

	"github.com/vmware-labs/wasm-workers-server/kits/go/worker"

	"github.com/tidwall/sjson"
)

func main() {
	worker.ServeFunc(func(w http.ResponseWriter, r *http.Request) {
		// Build a JSON body
		body, _ := sjson.Set("", "title", "New POST!")
		body, _ = sjson.Set(body, "body", "This is the body")
		body, _ = sjson.Set(body, "userId", 1)

		url := "https://jsonplaceholder.typicode.com/posts"

		// Create the request
		req, err := http.NewRequest(http.MethodPost, url, bytes.NewBufferString(body))
		if err != nil {
			panic(err)
		}
		req.Header.Set("Content-Type", "application/json")

		res, err := worker.SendHttpRequest(req)
		if err != nil {
			panic(err)
		}

		// Read the response
		resBody, err := io.ReadAll(res.Body)
		if err != nil {
			panic(err)
		}
		res.Body.Close()

		w.Header().Set("x-generated-by", "wasm-workers-server")
		w.Write([]byte(resBody))
	})
}
```

File: docs/docs/languages/go.md [hunk 26]
```diff
@@ -368,7 +381,7 @@ You can define [dynamic routes by adding route parameters to your worker files](
         "fmt"
         "net/http"
 
-        "github.com/vmware-labs/wasm-workers-server/kits/go/worker"
+i"github.com/vmware-labs/wasm-workers-server/kits/go/worker"
```
Comment (ereslibre) [inline docs/docs/languages/go.md:384]: Typo here :)

Comment (Angelmmiguel) [reply docs/docs/languages/go.md]: Done!

**Review by ereslibre — APPROVED** [review]
LGTM! 🚢 it!

**Pull request merged**